### PR TITLE
DAOS-2390 container: Epoch code cleanup & online aggregation

### PR DESCRIPTION
**Pull request by NiuYawei**
This patch removes redundant code left over from the old epoch model,
due to changes to the DAOS API that resulted in epoch implementation
becoming transparent to the end user. It removes LRE, LHE and HCE
along with their global versions. Also fixes aggregation and snapshots
along with corresponding unit tests.

It also enabled online aggregation to coalesce committed records and
condense VOS trees. The aggregate epoch range is confined by existing
snapshots, and the upper bound is capped by 'current hlc - delta'
for this moment.

Snapshot creation has two phases. In phase 1, the service leader
broadcasts to all targets preventing them from aggregating beyond the
instantaneous HLC at each target. The broadcast also allows CaRT to
synchronize HLCs to the highest value; the sybnchronized HLC is chosen
as the value of the new snapshot and persisted into RDB. In phase 2,
the updated list of snapshots is relayed to all targets via IVs
allowing aggregation to resume normally. At this point, epochs falling
under any deleted snapshots are also aggregated away.

Limitations (To be fixed in another patch):
1. GHCE is retained for calculating an upper bound on aggregation
2. No IO barrier when creating snapshot, so concurrent updates with
   lower epoch could be committed after snapshot creation is done
3. src/tests/suite/daos_epoch.c needs be rewritten to test online
   aggregation and snapshots
4. Online aggregation is disabled for rebuild tests till 'stable
   view rebuild' is done

Signed-off-by: Omkar Kulkarni <omkar.kulkarni@intel.com>
Signed-off-by: Niu Yawei <yawei.niu@intel.com>
Change-Id: I26efa2e6428487deedd4b8336e7c8392d2b9bdce

### CODE REVIEW

File: src/container/srv_target.c [hunk 1]
```diff
@@ -848,6 +874,7 @@ ds_cont_tgt_open(uuid_t pool_uuid, uuid_t cont_hdl_uuid,
 		D_GOTO(out, rc);
 
 	cont->sc_iv_ns = pool->sp_iv_ns;
+	ds_cont_tgt_snapshots_refresh(pool_uuid, cont_uuid);
```
[the remark this answers is not in the document]
Comment (NiuYawei) [reply src/container/srv_target.c]: Sorry, missed this comment.

You are right, there is a race of snapshot IV refresh vs. aggregation ULT, I'll fix it.

Right, current implementation is that aggregation ULT start on container open, and it'll stop on container destroy (or free container cache), that's similar to dtx resync ULT. I think we may need to cleanup it a bit in the future.

**Comment by NiuYawei** [comment]
The patch passed CI testing finally, could you review it asp? Thanks.

File: src/vos/vos_aggregate.c [hunk 2]
```diff
@@ -1536,17 +1535,8 @@ static int
 aggregate_enter(struct vos_container *cont, bool discard)
 {
 	if (cont->vc_in_aggregation) {
-		D_ERROR(DF_CONT": Already in ggregation. discard:%d\n",
+		D_ERROR(DF_CONT": Already in agregation. discard:%d\n",
```
Comment (jolivier23) [inline src/vos/vos_aggregate.c]: correctly added the a but there are 2 g's in aggregation :-)

File: src/tests/suite/daos_obj.c [hunk 3]
```diff
@@ -2247,6 +2249,13 @@ close_reopen_coh_oh(test_arg_t *arg, struct ioreq *req, daos_obj_id_t oid)
 static void
 tx_discard(void **state)
 {
+	/*
+	 * FIXME: This obsolete epoch model transaction API test have been
+	 * broken by online aggregation, needs be removed or updated as per
+	 * new transaction model.
+	 */
+	print_message("Skip obsolete test\n");
```
Comment (jolivier23) [inline src/tests/suite/daos_obj.c]: should use skip(); ?  That way skipped tests are in report.

File: src/container/srv_target.c [hunk 4]
```diff
@@ -479,17 +527,18 @@ cont_child_destroy_one(void *vin)
...
+		if (cont->sc_dtx_resyncing)
 			ABT_cond_wait(cont->sc_dtx_resync_cond, cont->sc_mutex);
-		}
 		ABT_mutex_unlock(cont->sc_mutex);
 		/* Should evict if idle, but no such interface at the moment. */
 		cont_child_put(tls->dt_cont_cache, cont);
 
-		if (!resyncing) {
+		retry_cnt++;
+		if (retry_cnt > 2) {
```
[the remark this answers is not in the document]
Comment (NiuYawei) [reply src/container/srv_target.c]: Just give it one more chance.

File: src/container/srv_target.c [hunk 5]
```diff
@@ -1159,146 +1221,309 @@ ds_cont_tgt_epoch_discard_aggregator(crt_rpc_t *source, crt_rpc_t *result,
...
-		goto pool_child;
+	/* Snapshot deleted, reset HAE */
+	if (cont->sc_snapshots_nr > args->snap_count) {
+		rc = vos_cont_ctl(cont->sc_hdl, VOS_CO_CTL_RESET_HAE);
+		if (rc)
+			D_ERROR(DF_UUID": Reset HAE failed. %d\n",
+				DP_UUID(cont->sc_uuid), rc);
 	}
+	cont->sc_snapshots_nr = args->snap_count;
+	cont->sc_aggregation_max = DAOS_EPOCH_MAX;
```
[the remark this answers is not in the document]
Comment (NiuYawei) [reply src/container/srv_target.c]: so to remove the cap of aggregation epoch which introduced by ongoing snapshot creation.

File: src/container/srv_target.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (Nasf-Fan) [reply src/container/srv_target.c]: If without CPU yield, more retry may help nothing, right?

File: src/container/srv_target.c [hunk 6]
```diff
@@ -1159,146 +1221,309 @@ ds_cont_tgt_epoch_discard_aggregator(crt_rpc_t *source, crt_rpc_t *result,
...
-	out->tao_rc = (rc == 0 ? 0 : 1);
-	D_DEBUG(DF_DSMS, DF_CONT": replying rpc %p: %d (%d)\n",
-		DP_CONT(in->tai_pool_uuid, in->tai_cont_uuid),
-		rpc, out->tao_rc, rc);
+	out->tao_rc = 0;
+	D_DEBUG(DF_DSMS, DF_CONT": replying rpc %p: %d\n",
+		DP_CONT(in->tai_pool_uuid, in->tai_cont_uuid), rpc,
+		out->tao_rc);
 	crt_reply_send(rpc);
 	if (out->tao_rc != 0)
```
[the remark this answers is not in the document]
Comment (NiuYawei) [reply src/container/srv_target.c]: Right, these are all dead code to be removed or revised.

File: src/container/srv_target.c [hunk 7]
```diff
@@ -1159,146 +1221,309 @@ ds_cont_tgt_epoch_discard_aggregator(crt_rpc_t *source, crt_rpc_t *result,
...
+	 * transactions under this epoch is either committed or aborted).
+	 */
+	epoch_max = hlc - interval;
+
+	/* Throttle the aggregation a bit */
+	if (cinfo.ci_hae > epoch_max - interval)
+		return 0;
+
+	/* Cap the aggregation upper bound to the snapshot in creating */
+	D_ASSERT(cont->sc_aggregation_max > 0);
```
[the remark this answers is not in the document]
Comment (NiuYawei) [reply src/container/srv_target.c]: Indeed, it's leftover from prior version.

**Pull request merged**